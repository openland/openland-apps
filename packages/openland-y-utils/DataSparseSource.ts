export type KeyExtractor<T> = (item: T) => string;
export type SeqExtractor<T> = (item: T) => number;

function isUnknown(v: number) {
    return Number.isNaN(v);
}

interface Mutator<T> {
    prepend(item: T): void;
    append(item: T): void;
    // loaded(items: T[],beginEnded: ): void;
}

export class DataSparseSource<T> {
    static readonly UNKNOWN = Number.NaN;

    readonly keyExtractor: KeyExtractor<T>;
    readonly seqExtractor: SeqExtractor<T>;
    readonly mutator: Mutator<T>;

    //
    // Numbering of elements is in ascending order:
    // 
    // | 1 <- first/begin
    // | 2
    // | 3 <- firstWindowCursor
    // | <empty>
    // ....
    // | <empty>
    // | 1234 <- lastWindowCursor
    // | 1235
    // | 1236 <- last/end
    //

    // Maximum id in first window
    private _completed = false;
    private _firstWindowCursor: number = DataSparseSource.UNKNOWN;
    private _firstWindow: T[] = [];
    // Minimum id in last window
    private _lastWindowCursor: number = DataSparseSource.UNKNOWN;
    private _lastWindow: T[] = [];
    // All windows except first and last one
    private _windows: { start: number, end: number, items: T[] }[] = [];
    // All items
    private _items = new Map<string, { item: T, deleted: boolean }>();

    constructor(extractors: { key: KeyExtractor<T>, seq: SeqExtractor<T> }) {
        this.keyExtractor = extractors.key;
        this.seqExtractor = extractors.seq;
        this.mutator = {
            prepend: this._prepend,
            append: this._append
        };
    }

    //
    // Mutations
    //

    private _append = (item: T): void => {
        let seq = this.seqExtractor(item);
        let key = this.keyExtractor(item);

        // Internal key value store
        if (this._items.has(key)) {
            throw Error('Item ' + key + ' already exists');
        }
        this._items.set(key, { item: item, deleted: false });

        // Check if merging required
        if (!this._completed && !isUnknown(seq)) {
            // If seq is known we should check for merge
            if ((isUnknown(this._lastWindowCursor) || seq < this._lastWindowCursor)) {
                this._lastWindowCursor = seq;
                // If lastWindowCursor is unknown or seq is bigger then we should check for merge

                if (!isUnknown(this._firstWindowCursor) && seq <= this._firstWindowCursor) {
                    // seq is smaller than first window seq - merge needed
                    this._lastWindow = [...this._firstWindow, ...this._lastWindow];
                    this._completed = true;
                }
            }
        }

        // Append item
        this._lastWindow.push(item);
    }

    private _prepend = (item: T): void => {
        let seq = this.seqExtractor(item);
        let key = this.keyExtractor(item);

        // Internal key value store
        if (this._items.has(key)) {
            throw Error('Item ' + key + ' already exists');
        }
        this._items.set(key, { item: item, deleted: false });

        // Check if merging required
        if (!this._completed && !isUnknown(seq)) {
            // If seq is known we should check for merge
            if ((isUnknown(this._firstWindowCursor) || this._firstWindowCursor < seq)) {
                this._firstWindowCursor = seq;
                // If _firstWindowCursor is unknown or seq is smaller then we should check for merge

                if (!isUnknown(this._lastWindowCursor) && this._lastWindowCursor <= seq) {
                    // seq is smaller than first window seq - merge needed
                    this._lastWindow = [...this._firstWindow, ...this._lastWindow];
                    this._completed = true;
                }
            }
        }

        // Append item
        if (this._completed) {
            this._lastWindow.unshift(item);
        } else {
            this._firstWindow.unshift(item);
        }
    }

    //
    // Dump
    //

    dump = () => {
        if (this._completed) {
            return {
                last: {
                    items: [...this._lastWindow]
                }
            };
        }
        let res: any = {};
        if (this._firstWindow.length > 0) {
            res.first = {
                end: this._firstWindowCursor,
                items: [...this._firstWindow]
            };
        }
        if (this._windows.length !== 0) {
            res.windows = [];
            for (let w of this._windows) {
                res.windows.push({
                    start: w.start,
                    end: w.end,
                    items: [...w.items]
                });
            }
        }
        if (this._lastWindow.length > 0) {
            res.last = {
                start: this._lastWindowCursor,
                items: [...this._lastWindow]
            };
        }
        return res;
    }
}