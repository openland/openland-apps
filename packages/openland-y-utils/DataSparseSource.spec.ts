import { DataSparseSource } from './DataSparseSource';

function createDataSource() {
    return new DataSparseSource<{ key: string, seq: number }>({
        key: (i) => i.key,
        seq: (i) => i.seq
    });
}

describe('DataSparseSource', () => {
    it('should append and prepend new items', () => {
        let ds = createDataSource();
        ds.mutator.append({ key: '1', seq: DataSparseSource.UNKNOWN });
        ds.mutator.append({ key: '2', seq: 1 });
        ds.mutator.append({ key: '3', seq: 2 });
        ds.mutator.append({ key: '4', seq: 3 });
        expect(ds.dump()).toMatchSnapshot();

        ds.mutator.prepend({ key: '5', seq: DataSparseSource.UNKNOWN });
        ds.mutator.prepend({ key: '6', seq: -1 });
        ds.mutator.prepend({ key: '7', seq: -2 });
        ds.mutator.prepend({ key: '8', seq: -3 });
        expect(ds.dump()).toMatchSnapshot();

        // Weird input should lead to merging
        ds.mutator.append({ key: '9', seq: -4 });
        expect(ds.dump()).toMatchSnapshot();

        expect(() => ds.mutator.append({ key: '4', seq: 3 })).toThrowError('Item 4 already exists');
        expect(() => ds.mutator.prepend({ key: '4', seq: 3 })).toThrowError('Item 4 already exists');
    });

    it('should merge prepend new items', () => {
        let ds = createDataSource();
        ds.mutator.append({ key: '1', seq: DataSparseSource.UNKNOWN });
        ds.mutator.append({ key: '2', seq: 1 });
        ds.mutator.append({ key: '3', seq: 2 });
        ds.mutator.append({ key: '4', seq: 3 });
        expect(ds.dump()).toMatchSnapshot();

        ds.mutator.prepend({ key: '5', seq: DataSparseSource.UNKNOWN });
        ds.mutator.prepend({ key: '6', seq: -1 });
        ds.mutator.prepend({ key: '7', seq: -2 });
        ds.mutator.prepend({ key: '8', seq: -3 });
        expect(ds.dump()).toMatchSnapshot();

        // Weird input should lead to merging
        ds.mutator.prepend({ key: '9', seq: 4 });
        expect(ds.dump()).toMatchSnapshot();
    });
});