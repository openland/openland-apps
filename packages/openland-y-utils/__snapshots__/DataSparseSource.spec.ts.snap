// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSparseSource should append and prepend new items 1`] = `
Object {
  "last": Object {
    "items": Array [
      Object {
        "key": "1",
        "seq": NaN,
      },
      Object {
        "key": "2",
        "seq": 1,
      },
      Object {
        "key": "3",
        "seq": 2,
      },
      Object {
        "key": "4",
        "seq": 3,
      },
    ],
    "start": 1,
  },
}
`;

exports[`DataSparseSource should append and prepend new items 2`] = `
Object {
  "first": Object {
    "end": -1,
    "items": Array [
      Object {
        "key": "8",
        "seq": -3,
      },
      Object {
        "key": "7",
        "seq": -2,
      },
      Object {
        "key": "6",
        "seq": -1,
      },
      Object {
        "key": "5",
        "seq": NaN,
      },
    ],
  },
  "last": Object {
    "items": Array [
      Object {
        "key": "1",
        "seq": NaN,
      },
      Object {
        "key": "2",
        "seq": 1,
      },
      Object {
        "key": "3",
        "seq": 2,
      },
      Object {
        "key": "4",
        "seq": 3,
      },
    ],
    "start": 1,
  },
}
`;

exports[`DataSparseSource should append and prepend new items 3`] = `
Object {
  "last": Object {
    "items": Array [
      Object {
        "key": "8",
        "seq": -3,
      },
      Object {
        "key": "7",
        "seq": -2,
      },
      Object {
        "key": "6",
        "seq": -1,
      },
      Object {
        "key": "5",
        "seq": NaN,
      },
      Object {
        "key": "1",
        "seq": NaN,
      },
      Object {
        "key": "2",
        "seq": 1,
      },
      Object {
        "key": "3",
        "seq": 2,
      },
      Object {
        "key": "4",
        "seq": 3,
      },
      Object {
        "key": "9",
        "seq": -4,
      },
    ],
  },
}
`;

exports[`DataSparseSource should merge prepend new items 1`] = `
Object {
  "last": Object {
    "items": Array [
      Object {
        "key": "1",
        "seq": NaN,
      },
      Object {
        "key": "2",
        "seq": 1,
      },
      Object {
        "key": "3",
        "seq": 2,
      },
      Object {
        "key": "4",
        "seq": 3,
      },
    ],
    "start": 1,
  },
}
`;

exports[`DataSparseSource should merge prepend new items 2`] = `
Object {
  "first": Object {
    "end": -1,
    "items": Array [
      Object {
        "key": "8",
        "seq": -3,
      },
      Object {
        "key": "7",
        "seq": -2,
      },
      Object {
        "key": "6",
        "seq": -1,
      },
      Object {
        "key": "5",
        "seq": NaN,
      },
    ],
  },
  "last": Object {
    "items": Array [
      Object {
        "key": "1",
        "seq": NaN,
      },
      Object {
        "key": "2",
        "seq": 1,
      },
      Object {
        "key": "3",
        "seq": 2,
      },
      Object {
        "key": "4",
        "seq": 3,
      },
    ],
    "start": 1,
  },
}
`;

exports[`DataSparseSource should merge prepend new items 3`] = `
Object {
  "last": Object {
    "items": Array [
      Object {
        "key": "9",
        "seq": 4,
      },
      Object {
        "key": "8",
        "seq": -3,
      },
      Object {
        "key": "7",
        "seq": -2,
      },
      Object {
        "key": "6",
        "seq": -1,
      },
      Object {
        "key": "5",
        "seq": NaN,
      },
      Object {
        "key": "1",
        "seq": NaN,
      },
      Object {
        "key": "2",
        "seq": 1,
      },
      Object {
        "key": "3",
        "seq": 2,
      },
      Object {
        "key": "4",
        "seq": 3,
      },
    ],
  },
}
`;
